proposals/12238-notebook-component/README.md

```diff
@@ -0,0 +1,295 @@
+currently requires boilerplate: exporting to Python, writing a wrapper function, or managing custom container images. A
+native notebook component:
+
+- Reduces friction to productionize notebook code in pipelines
+- Preserves the notebook as the source of truth while allowing parameterization
+- Avoids extra build steps by embedding notebook content into the component
+
+### Goals
+
+1. Enable defining a component from a `.ipynb` notebook with a single decorator.
+2. Support parameter injection into the notebook at execution time.
+3. Use the existing Python executor to keep compatibility with existing inputs/outputs concepts.
+
+### Non-Goals
+
+1. Notebook validation or linting beyond JSON and structural checks.
+2. Adding a new pipeline IR or backend executor. This reuses the Python executor.
+
+## Proposal
+
+### Baseline Feature For Bundled Assets
+
+This KEP establishes a baseline, generic capability for Python-function components to embed arbitrary files or
+directories directly into a lightweight component. The goal is to support cases where a component needs a small amount
+of read-only assets (configs, scripts, models, notebooks, etc.) without requiring a custom image.
+
+- Add a new decorator argument to `@dsl.component`:
+  - `bundled_artifact_path: Optional[str]` — path to a file or directory on the authoring machine to bundle with the
+    component.
+- Add a new SDK type: `dsl.BundledInput[T]` — a runtime-only input annotation that resolves to an artifact instance of
+  type `T` (e.g. `dsl.Dataset`) whose `path` points to the extracted bundle root.
+  - If a directory is bundled, `.path` points to the extracted directory.
+  - If a single file is bundled, `.path` points to that file.
+
+Example:
+
+```python
+@dsl.component(
+    bundled_artifact_path="assets/config_dir",
+)
+def my_component(cfg: dsl.BundledInput[dsl.Dataset], param: int):
+    # cfg.path is a directory when a directory is bundled; a file path if a file is bundled
+    print(cfg.path)
+```
+
+Execution model (lightweight components):
+
+- At compile time, the file/dir is archived (tar + gzip) and base64-embedded into the ephemeral module.
+- At runtime, the bundle is safely extracted to a temporary location and the `BundledInput[...]` parameter is injected
+  as an artifact with `.path` pointing to the extracted file/dir in a temporary directory.
+
+Relationship to `@dsl.notebook_component`:
+
+- The notebook component leverages the same embedding pattern, specializing the runtime helper to execute `.ipynb`
+  content via `nbconvert`. If `notebook_path` is a directory, the `.ipynb` file in the directory is executed to allow
+  for utility functions to be bundled. Pipeline compilation fails if there is more than one `.ipynb` file.
+
+### SDK User Experience
+
+#### Example
+
+```python
+from kfp import dsl
+
+
+@dsl.notebook_component(
+    notebook_path="train.ipynb",
+    packages_to_install=["pandas", "scikit-learn", "nbconvert", "jupyter"],
```

Comment:
Maybe have an index URL as an optional parameter that defaults to public pypi? This allows using private/internal repositories

Reply:
Agreed. `notebook_component` is wrapping the `component` decorator, so it'll have the `pip_trusted_hosts` argument available:
https://github.com/kubeflow/pipelines/blob/ffaeeda3da4f68c6b39231828ab23ff5f2a367e7/sdk/python/kfp/dsl/component_decorator.py#L34